Tests/RouterTests.swift

```diff
@@ -43,14 +43,55 @@ private final class RouterTestServer: NSURLProtocol {
     override func startLoading() {
         guard let requestURL = request.URL,
                   client = client else { return }
-        
+
         client.URLProtocol(self, didReceiveResponse: NSHTTPURLResponse(URL: requestURL, statusCode: 200, HTTPVersion: "HTTP/1.1", headerFields: nil)!, cacheStoragePolicy: .AllowedInMemoryOnly)
         client.URLProtocolDidFinishLoading(self)
     }
 
     override func stopLoading() {}
 }
 
+private final class ProtocolClientTest: NSObject, NSURLProtocolClient {
+
+    private let shouldFail_URLProtocolDidFinishLoading: Bool
+    
+    var didCall_URLProtocolDidFinishLoading: Bool = false
+    
+    init(shouldFail_URLProtocolDidFinishLoading: Bool) {
+        self.shouldFail_URLProtocolDidFinishLoading = shouldFail_URLProtocolDidFinishLoading
+        super.init()
+    }
+
+    @objc func URLProtocol(`protocol`: NSURLProtocol, wasRedirectedToRequest request: NSURLRequest, redirectResponse: NSURLResponse) {
+        // intentionally left empty
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, cachedResponseIsValid cachedResponse: NSCachedURLResponse) {
+        // intentionally left empty
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, didReceiveResponse response: NSURLResponse, cacheStoragePolicy policy: NSURLCacheStoragePolicy) {
+        // intentionally left empty
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, didLoadData data: NSData) {
+        // intentionally left empty
+    }
+    @objc func URLProtocolDidFinishLoading(`protocol`: NSURLProtocol) {
+        if shouldFail_URLProtocolDidFinishLoading {
+            XCTFail("Protocol client should never get 'URLProtocolDidFinishLoading' called.")
+        }
+        
+        didCall_URLProtocolDidFinishLoading = true
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, didFailWithError error: NSError) {
+        // intentionally left empty
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, didReceiveAuthenticationChallenge challenge: NSURLAuthenticationChallenge) {
+        // intentionally left empty
+    }
+    @objc func URLProtocol(`protocol`: NSURLProtocol, didCancelAuthenticationChallenge challenge: NSURLAuthenticationChallenge) {
+        // intentionally left empty
+    }
+}
+
```

Comment:
What is this `ProtocolClientTest` for? I think we shouldn't need to create it. 

Basically, the normal way to test it would be:

1. With any implementation, add the test that cancels the data task and expects the request to not get started. Thus, this test will fail.
2. Add the implementation to stop the requests in KakapoServer
3. Check the test again, should pass.

For the test itself: should be enough to get the task from `dataTaskWithURL` and cancel it, and expect the request to not being called. Same as you do in `"should not start a request when request gets cancelled"`. We could also add other test with more than one request or with one canceled request and another not cancelled request, but that should be it. I don't think we need more tests with a new `NSURLProtocolClient`